### PR TITLE
Replace non-portable calls to Fortran BLAS and LAPACK with calls to CBLAS and LAPACKE, respectively [feature/use-cblas-and-lapacke] 

### LINKED ISSUE #397: Issue with Fortran name mangling on BG/Q - LAPACK
I'm trying to compile MFEM (commit bc6b484, tag v3.3.2) on LLNL's Vulcan - a BlueGene/Q machine.
I'm trying to use the system provided `ESSL/LAPACK/BLAS` libraries and I'm running into some problems with undefined references when I try to compile the examples.

Configuration:
```
  make config \
  HYPRE_OPT=-I/g/g14/kudo4/vulcan_install/hypre/include \
  HYPRE_LIB="-L/g/g14/kudo4/vulcan_install/hypre/lib -lHYPRE" \
  METIS_OPT=-I/g/g14/kudo4/vulcan_install/metis_debug/include \
  METIS_LIB="-L/g/g14/kudo4/vulcan_install/metis_debug/lib -lmetis" \
  NETCDF_OPT=-I/usr/local/tools/netcdf-4.1.2/include/ \
  NETCDF_LIB="-L/usr/local/tools/netcdf-4.1.2/lib -lnetcdf -L/usr/local/tools/curl-7.14.1/lib -lcurl -L/usr/local/tools/zlib-1.2.6/lib -lz" \
  BLAS_LIB="/usr/local/tools/essl/5.1/lib/libesslsmpbg.a /usr/local/tools/blas/lib/libblas.a -L/opt/ibmcmp/xlf/bg/14.1/bglib64/ -L/opt/ibmcmp/xlsmp/bg/3.1/bglib64/ -lxlf90_r -lxlsmp -lxlopt -lxlfmath -lxl" \
  LAPACK_LIB="/usr/local/tools/essl/5.1/lib/libesslsmpbg.a /usr/local/tools/lapack/lib/liblapack.a /usr/local/tools/essl/5.1/lib/libesslsmpbg.a /usr/local/tools/blas/lib/libblas.a -L/opt/ibmcmp/xlf/bg/14.1/bglib64/ -L/opt/ibmcmp/xlsmp/b\
g/3.1/bglib64/ -lxlf90_r -lxlsmp -lxlopt -lxlfmath -lxl" \
  CXX=/usr/local/bin/mpixlcxx_r-fastmpi  \
  MPICXX=/usr/local/bin/mpixlcxx_r-fastmpi  \
  PREFIX=/g/g14/kudo4/vulcan_install/mfem \
  CXXFLAGS="-O0 -g" \
  MFEM_USE_MPI=YES \
  MFEM_USE_METIS_5=NO \
  MFEM_DEBUG=NO \
  MFEM_USE_GZSTREAM=NO \
  MFEM_USE_LIBUNWIND=NO \
  MFEM_USE_LAPACK=YES \
  MFEM_THREAD_SAFE=NO \
  MFEM_USE_OPENMP=NO \
  MFEM_USE_MEMALLOC=YES \
  MFEM_TIMER_TYPE=2 \
  MFEM_USE_SUNDIALS=NO \
  MFEM_USE_MESQUITE=NO \
  MFEM_USE_SUITESPARSE=NO \
  MFEM_USE_ARPACK=NO \
  MFEM_USE_SUPERLU=NO \
  MFEM_USE_GECKO=NO \
  MFEM_USE_GNUTLS=NO \
  MFEM_USE_NETCDF=YES \
  MFEM_USE_MPFR=NO \
  MFEM_USE_SIDRE=NO
```

After I compile `libmfem.a`, I try `make ex1` in `/examples`:
```
/usr/local/bin/mpixlcxx_r-fastmpi  -O0 -g -I.. -I/g/g14/kudo4/vulcan_install/hypre/include -I/g/g14/kudo4/vulcan_install/metis_debug/include  -I/usr/local/tools/netcdf-4.1.2/include/ ex1.cpp -o ex1 -L.. -lmfem -L/g/g14/kudo4/vulcan_install/hypre/lib -lHYPRE -L/g/g14/kudo4/vulcan_install/metis_debug/lib -lmetis /usr/local/tools/essl/5.1/lib/libesslsmpbg.a /usr/local/tools/lapack/lib/liblapack.a /usr/local/tools/essl/5.1/lib/libesslsmpbg.a -L/opt/ibmcmp/xlf/bg/14.1/bglib64/ -L/opt/ibmcmp/xlsmp/bg/3.1/bglib64/ -lxlf90_r -lxlsmp -lxlopt -lxlfmath -lxl -L/usr/local/tools/netcdf-4.1.2/lib -lnetcdf -L/usr/local/tools/curl-7.14.1/lib -lcurl -L/usr/local/tools/zlib-1.2.6/lib -lz -lrt
/usr/local/tools/curl-7.14.1/lib/libcurl.a(ldap.o): In function `DynaOpen':
/g/g0/rhedges/SEQ-netcdf/curl-7.14.1/lib/ldap.c:129: warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
../libmfem.a(densemat.o): In function `mfem::DenseMatrixSVD::Eval(mfem::DenseMatrix&)':
/g/g14/kudo4/vulcan_src/mfem/linalg/densemat.cpp:4303: undefined reference to `dgesvd_'
../libmfem.a(densemat.o): In function `mfem::DenseMatrixSVD::Init()':
/g/g14/kudo4/vulcan_src/mfem/linalg/densemat.cpp:4283: undefined reference to `dgesvd_'
../libmfem.a(densemat.o): In function `mfem::DenseMatrixEigensystem::Eval()':
/g/g14/kudo4/vulcan_src/mfem/linalg/densemat.cpp:4237: undefined reference to `dsyev_'
../libmfem.a(densemat.o): In function `mfem::DenseMatrixEigensystem::DenseMatrixEigensystem(mfem::DenseMatrix&)':
/g/g14/kudo4/vulcan_src/mfem/linalg/densemat.cpp:4218: undefined reference to `dsyev_'
../libmfem.a(densemat.o): In function `mfem::DenseMatrixEigensystem::DenseMatrixEigensystem(mfem::DenseMatrix&)':
/g/g14/kudo4/vulcan_src/mfem/linalg/densemat.cpp:4218: undefined reference to `dsyev_'
../libmfem.a(densemat.o): In function `mfem::DenseMatrix::SingularValues(mfem::Vector&) const':
/g/g14/kudo4/vulcan_src/mfem/linalg/densemat.cpp:1110: undefined reference to `dgesvd_'
/g/g14/kudo4/vulcan_src/mfem/linalg/densemat.cpp:1116: undefined reference to `dgesvd_'
../libmfem.a(densemat.o): In function `mfem::dsyev_Eigensystem(mfem::DenseMatrix&, mfem::Vector&, mfem::DenseMatrix*)':
/g/g14/kudo4/vulcan_src/mfem/linalg/densemat.cpp:1058: undefined reference to `dsyev_'
/g/g14/kudo4/vulcan_src/mfem/linalg/densemat.cpp:1063: undefined reference to `dsyev_'
../libmfem.a(densemat.o): In function `mfem::dsyevr_Eigensystem(mfem::DenseMatrix&, mfem::Vector&, mfem::DenseMatrix*)':
/g/g14/kudo4/vulcan_src/mfem/linalg/densemat.cpp:907: undefined reference to `dsyevr_'
/g/g14/kudo4/vulcan_src/mfem/linalg/densemat.cpp:917: undefined reference to `dsyevr_'
```

If I take a look at the symbols defined in the system provided `liblapack.a` via `nm --defined-only /usr/local/tools/lapack/lib/liblapack.a`, I see symbols along the lines of:
```
dgesvd.o:
0000000000000000 r .const_dr
0000000000000000 D dgesvd
```

So the problem seems to be that MFEM is assuming that Fortran global symbols contain underscores whereas the LAPACK library I'm trying to use does not append any underscores.  If I hand modify `/linalg/densemat.cpp` to reference the same symbols without the underscores,
```
--- a/linalg/densemat.cpp
+++ b/linalg/densemat.cpp
...
-dgetrf_(int *, int *, double *, int *, int *, int *);
+dgetrf(int *, int *, double *, int *, int *, int *);
...
```
I can successfully `make ex1` and execute it.

Is there a way to make MFEM naturally handle this situation during the build?

### PR DESCRIPTION
**Pull request by goxberry**
Fixes #397, #1002. 

# Rationale

As discussed in #1002, this set of patches is the simplest way to achieve portability of BLAS and LAPACK calls because it sidesteps the Fortran name-mangling issue by replacing calls to Fortran with calls to the C language interfaces to BLAS and LAPACK.

# Alternatives

- *Preserve the status quo:* As @v-dobrev [notes](https://github.com/mfem/mfem/issues/397#issuecomment-353158665) in #397, most MFEM use cases do not require LAPACK, and it could be argued that there is no reason to support LAPACK in MFEM on uncommon OS/compiler/processor architecture configurations. This decision is the prerogative of the maintainers, but such a decision should probably be noted in documentation. It also complicates a common development workflow: porting non-scalable code on a laptop or workstation to an HPC machine like a BG/Q for debugging, and as a starting point for replacing non-scalable algorithms with scalable ones like AMG or BLOPEX. A development workflow in this spirit led to #397.

- *Attempt to generate portable macros for Fortran name mangling:* This PR avoids tackling the general issue of portably name mangling Fortran function and subroutine calls for simplicity and expediency. If MFEM plans on using more Fortran APIs, portable name mangling macros would be preferable.

# Incompatibilities

This PR requires LAPACK 3.4.0 and BLAS 3.4.0, because LAPACKE first appears in LAPACK 3.4.0. Notably, merging this change into MFEM will mean that MFEM _cannot_ link to the macOS Accelerate framework, which is Apple's vendor BLAS/LAPACK library. Accelerate is based on LAPACK 3.2.1, and appears to be unmaintained, which is [why libraries like SciPy have dropped support for Accelerate](https://github.com/scipy/scipy/wiki/Dropping-support-for-Accelerate), and [Homebrew formulas now link to OpenBLAS instead of Accelerate plus `veclibfort`](https://discourse.brew.sh/t/is-there-a-reason-math-formulas-use-veclibfort-instead-of-openblas/4658). This change will also require changes to the [suggested macOS `packages.yaml` configuration file for Spack](https://ceed.exascaleproject.org/spack/ceed1-darwin-x86_64-packages.yaml) discussed on the [CEED 1.0 Software Distribution website](https://ceed.exascaleproject.org/ceed-1.0/). Notably, Accelerate uses the C language LAPACK interface in CLAPACK, which is obsolete and essentially the output of `f2c` applied to LAPACK. I do not see a compelling reason to support CLAPACK.

In the event that a vendor BLAS/LAPACK library implements LAPACK 3.4.0 or later, but does not include LAPACKE, this PR will be incompatible with such a library. In this case, [LAPACKE could be copied from reference LAPACK](https://sourceforge.net/p/math-atlas/mailman/message/30690449/) to supply the necessary functionality. To the best of my knowledge, the following BLAS/LAPACK implementations do _not_ ship LAPACKE:

- Apple Accelerate (see above)
- ATLAS

The following BLAS/LAPACK library implementations do ship LAPACKE:

- Reference LAPACK
- OpenBLAS
- Intel MKL
- BLIS/libFLAME
- ACML (AMD)
- ESSL (IBM)
- libsci (Cray)

In other words, nearly all of the major BLAS/LAPACK implementations ship LAPACKE, so using it should not be an issue.

### CODE REVIEW

**Comment by goxberry** [comment]
cc: @junkudo Hopefully, this PR fixes the BG/Q bugs you ran into earlier.

**Comment by goxberry** [comment]
cc: @rw-anderson I plan on writing a similar PR for [libROM](https://github.com/LLNL/libROM) because there's no clean way to generate identical Fortran name mangling macros in CMake and GNU Autotools. The SVD code in MFEM may also be useful for libROM examples.

File: linalg/densemat.cpp [hunk 1]
```diff
@@ -4089,10 +4039,14 @@ void LUFactors::USolve(int m, int n, double *X) const
 void LUFactors::Solve(int m, int n, double *X) const
 {
 #ifdef MFEM_USE_LAPACK
+   int layout = LAPACK_COL_MAJOR;
```
Comment (goxberry) [inline linalg/densemat.cpp]: `layout` could be a member of `LUFactors`, depending on maintainer preferences.

File: linalg/densemat.cpp [hunk 2]
```diff
@@ -4326,15 +4280,10 @@ DenseMatrixEigensystem::DenseMatrixEigensystem(DenseMatrix &m)
    ev.SetDataAndSize(NULL, n);
 
 #ifdef MFEM_USE_LAPACK
+   int layout = LAPACK_COL_MAJOR;
```
Comment (goxberry) [inline linalg/densemat.cpp]: `layout` could be a member of `DenseMatrixEigensystem`, depending on maintainer preferences.

**Comment by goxberry** [comment]
As an overall note: I've chosen to use the "high-level" LAPACKE interfaces, which automatically allocate work arrays inside the API calls. If the MFEM maintainers would prefer computing work array sizes and allocating work array space manually, the high-level LAPACKE calls could be replaced with "low-level" LAPACKE calls that require allocated work arrays and work array sizes.
